pkg/cli/image/mirror/mirror.go

```diff
@@ -179,14 +179,16 @@ func NewCmdMirrorImage(streams genericclioptions.IOStreams) *cobra.Command {
 }
 
 func (o *MirrorImageOptions) Complete(cmd *cobra.Command, args []string) error {
+	// Check before completing filter options, to see if filter-by-os flag is set
+	// Return manifest list if a user is mirroring a particular os/arch (or wildcard os/arches)
+	// The default is to mirror the single manifest that matches the user's system os/arch
+	if len(o.FilterOptions.FilterByOS) > 0 {
```

Comment:
see my other comment - I think omitting this makes the change less disruptive to existing oc image mirror users.

Reply:
I think maybe all that's needed for this PR is a validation that one cannot pass both `--filter-by-os=os/arch` (except wildcard) _and_ `--keep-manifest-list=true`  since they contradict.  wdyt?

Reply:
@ecordell ptal, I've updated.  If a user passes both `--filter-by-os=not/wildcard` _and_ `--keep-manifest-list=true`, they'll see a warning during the flag completion, then, the cmd will fail later on when it is determined that len(filtered) != len(list) - bc at that point the manifestList digest will be changed to a new one that includes only the filtered os/arch manifests.  Result will be like so:
```console
$ oc image  mirror quay.io/prometheus/prometheus:v2.18.1=localhost:5000/prom-part --keep-manifest-list=true --filter-by-os=linux/arm*
W1201 13:43:53.332411   93410 mirror.go:186] This command will fail if the filter ( linux/arm* ) does not include all available os/arch options.

error: failing because manifest list cannot be preserved with the given filters. Try running the command again without the --filter-by-os or to filter the list, omit the '--keep-manifest-list=true'. A filtered manifest list will have a new digest.
error: an error occurred during planning
```

Reply:
shouldn't that be allowed? 

--filter-by-os=os/arch and --keep-manifest-list: mirror only the manifests that match the filter in the whole manifestlist, mirror the manifestlist unmodified

--filter-by-os=os/arch: mirror only the manifests that match the filter in the whole manifestlist, mirror a modified manifestlist that only references the mirrored manifests

Reply:
i discussed this w/ @nalind and @mtrmac on slack (pinged you) - it's not possible to preserve the manifest list digest if its constituents are changed, so unless the filter includes all available os/arch, the manifest list digest needs to be updated, otherwise as i understand it, registries will refuse it.

Reply:
Ah you updated while I was typing the last one 😄 

> bc at that point the manifestList digest will be changed to a new one that includes only the filtered os/arch manifests.

The issue is that we have disconnected users that want to only mirror a subset of the manifests (i.e. the s390x ones) but need the entire manifestlist preserved (because they're referenced by digest in the cluster). So the `--keep-manifest-list` was supposed to preserve the original manifestlist even if only a subset of manifests contained within were actually mirrored.

Reply:
Right, hm.. I see this is currently possible by passing no --filter-by-os and with --keep-manifest-list=true. It seems we are already bending the rule there (a user _can_ today successfully mirror only their system os/arch with the reference to the manifest list, and then pull their system os/arch manifest while also being able to  pull the manifest list digest.)  The issue (the bz) is that a user cannot extend that to pass --filter-by-os=anything/other than the system os/arch - let me look at this again and see if there's a solution, since we already accept that we'll mirror the manifest list with only a single constituent.  Looking...

Reply:
er, no, I was wrong above, the manifest list digest is not preserved now unless one passes the wildcard filter-by-os.  - I've been working with this image: 
`quay.io/prometheus/prometheus@sha256:5880ec936055fad18ccee798d2a63f64ed85bd28e8e0af17c6923a090b686c3d` that is a manifest list (quay.io/prometheus/prometheus:v2.18.1)
If I run oc image mirror cmds I see the following:
```console
$ oc image  mirror quay.io/prometheus/prometheus:v2.18.1=localhost:5000/prom:v2.18.1 --keep-manifest-list=true
I1202 14:47:28.599652  226774 manifest.go:404] Filtered manifest list to new digest sha256:debb32e52e5b585d84f35e7867f680c11871f9ad31aba422c8bb4d44c23aea32
(that is a new manifest list digest since it filtered my system os/arch only)
but...
$ oc image  mirror quay.io/prometheus/prometheus:v2.18.1=localhost:5000/prom:v2.18.1 --filter-by-os=.*
sha256:5880ec936055fad18ccee798d2a63f64ed85bd28e8e0af17c6923a090b686c3d localhost:5000/prom:v2.18.1  the digest is preserved  (adding --keep-manifest-list=true is unnecessary, does nothing, it's the fact that all components of the list are filtered, or len(filtered) == len(list), that result in preserving the manifest list.  
```
So, what I've uncovered is that `--keep-manifest-list=true` is broken :shrug:  and has never worked, AFAICT.

Reply:
and to "fix" this means introducing a hack, to force preserve the manifest list digest - even if we change the contents of the manifest list.

Reply:
After further consideration I think this PR makes sense, unfortunately - when a user passes `--keep-manifest-list=true` the cmd should error if that can't happen (it does not, now, see the first of the 2 example cmds, above).
I think, however, we need to fix that first command above, to _not_ default to the system os/arch if `--keep-manifest-list=true` and instead mirror all os/arches since user has signalled they _want_ to preserve the manifest list - and only way to do that is to mirror all.  
I don't believe there is a fix to `--keep-manifest-list=true` and also `--filter-by-os=anything/other-than-wildcard` and the fact that our code is currently advertising otherwise is a bug (that this PR fixes)

Reply:
I agree - not sure there's a better solution if registries reject manifestlists that reference manifests not available in the registry

Reply:
You can't change the behavior of the flags.  Keep-manifest-list works in some scenarios, and we can't change the flags because those flags may be used in mirroring automation.  That's why the wildcard default behavior was introduced.

Can you summarize what problem you are trying to solve right now? :)

Reply:
I don't want to change the behavior of the flags, but as/is, --keep-manifest-list=true is broken, so to fix it, whenever a user passes `--keep-manifest-list=true` the mirror command must mirror _all_ manifests in the list, otherwise the manifest list digest is not preserved.

Reply:
I've updated this PR, so now, the change is that whenever `--keep-manifest-list=true` is passed, `--filter-by-os=.*`  since this is the only way to include the same manifest list digest.  The help menu explains that `--keep-manifest-list=true` will override any value passed for `--filter-by-os=not/wildcard`